hri_safety_sense/src/VscProcess.cpp

```diff
@@ -103,6 +107,78 @@ VscProcess::~VscProcess()
 	if(joystickHandler) delete joystickHandler;
 }
 
+void VscProcess::receivedVibration(const std_msgs::Bool msg)
+{
+	bool received_msg = msg.data;
+	if (received_msg == true)
+	{
+		vsc_send_user_feedback(vscInterface, VSC_USER_BOTH_MOTOR_INTENSITY, MOTOR_CONTROL_INTENSITY_HIGH);
+	}
+}
+
+void VscProcess::receivedDisplayOnCommand(const hri_safety_sense::SrcDisplay& msg)
+{	
+	// Save the first message we receive as the previous message
+	static bool msg_received = false;
+	if (!msg_received)
+	{
+		prev_msg_ = msg;
+		msg_received = true;
+		vsc_send_user_feedback_string(vscInterface, VSC_USER_DISPLAY_ROW_1, msg.displayrow1.c_str());
+		vsc_send_user_feedback_string(vscInterface, VSC_USER_DISPLAY_ROW_2, msg.displayrow2.c_str());
+		vsc_send_user_feedback_string(vscInterface, VSC_USER_DISPLAY_ROW_3, msg.displayrow3.c_str());
+		vsc_send_user_feedback_string(vscInterface, VSC_USER_DISPLAY_ROW_4, msg.displayrow4.c_str());
+		return;
+	}
+
+	// Turn on custom display mode
+	vsc_send_user_feedback(vscInterface, VSC_USER_DISPLAY_MODE, DISPLAY_MODE_CUSTOM_TEXT);
+
+	// Check if display message is above MAXCHARACTERS for a row in SRC display
+	if(msg.displayrow1.size() > msg.MAXCHARACTERS)
+	{
+		ROS_WARN("Maximum characters limit reached for display row 1. Please enter upto 20 characters.");
+	}
+	if(msg.displayrow2.size() > msg.MAXCHARACTERS)
+	{
+		ROS_WARN("Maximum characters limit reached for display row 2. Please enter upto 20 characters.");
+	}
+	if(msg.displayrow3.size() > msg.MAXCHARACTERS)
+	{
+		ROS_WARN("Maximum characters limit reached for display row 3. Please enter upto 20 characters.");
+	}
+	if(msg.displayrow4.size() > msg.MAXCHARACTERS)
+	{
+		ROS_WARN("Maximum characters limit reached for display row 4. Please enter upto 20 characters.");
+	}
+	
+	// Only send a display message if it is different from the previous message
+	if(msg.displayrow1 != prev_msg_.displayrow1)
+	{
+		vsc_send_user_feedback_string(vscInterface, VSC_USER_DISPLAY_ROW_1, msg.displayrow1.c_str());
+	}
+	if(msg.displayrow2 != prev_msg_.displayrow2)
+	{
+		vsc_send_user_feedback_string(vscInterface, VSC_USER_DISPLAY_ROW_2, msg.displayrow2.c_str());
+	}
+	if(msg.displayrow3 != prev_msg_.displayrow3)
+	{
+		vsc_send_user_feedback_string(vscInterface, VSC_USER_DISPLAY_ROW_3, msg.displayrow3.c_str());
+	}
+	if(msg.displayrow4 != prev_msg_.displayrow4)
+	{
+		vsc_send_user_feedback_string(vscInterface, VSC_USER_DISPLAY_ROW_4, msg.displayrow4.c_str());
+	}
```

Comment:
If the display has been turned OFF (back to DISPLAY_MODE_STANDARD), do we need to re-send our previous strings, or does it save them internally?

Reply:
Once the display is put to DISPLAY_CUSTOM_MODE to OFF, to redisplay data, the display would need to be put in DISPLAY_CUSTOM_MODE to ON (`rostopic pub /src_display_mode_on hri_safety_sense/SrcDisplay "displayrow1: 'new_message'
displayrow2: ''
displayrow3: ''
displayrow4: ''"`) which will display a new message.

Do you want me to change and publish topics such as `/src_custom_display_mode_on` (turns on custom display), `/src_custom_display_mode_off` (turns off custom display) and `/src_custom_display` (displays custom display)?

Reply:
Adding new topics I think would complicate it. What we want is to just be able to send a message to display something, and it displays. My concern is that right now, if you Send a message like this:
```
Display
Some
Text
Here
```

And then you send a message to turn off the display, and then send the same message again, or a message that is somewhat the same, like this:
```
Display
My
Text
There
```
What you'll end up seeing is
```

My

Here
```

Because it will only update two of the 4 lines.

Reply:
Tried turning OFF the display and turning it back ON and it saves the previous messgae. It displays the last published message. @AGummyBear

Reply:
Thanks for checking. I think it would be good to add that line to clear the last message after turning off the display. You may need to adjust the functionality to make sure it displays the new message correctly when turning it back on. Basically, test that scenario that I suggested.

Reply:
Okay. I tried testing your scenario and I am able to get
```if you Send a message like this:

Display
Some
Text
Here

And then you send a message to turn off the display, and then send the same message again, or a message that is somewhat the same, like this:

Display
My
Text
There
What you'll end up seeing is ```

I am able to see 
``` Display
My
Text
There```

Reply:
Did you change it to clear out the display after turning it off?

Reply:
Not yet. I was testing your scenario.

Reply:
Yeah. I ended up seeing

My

Here

if I clear out the display after turning it off.

Reply:
Fixed this by storing clear message as previous message.